### ISSUE TITLE
files not receiving on android 11

### ISSUE THREAD

**Issue by devalanrs** [mannprerak2/nearby_connections#33]
<details><summary>Error Logs</summary>
D/nearby_connections(17721): ad.onConnectionResult
D/nearby_connections(17721): onPayloadReceived
D/nearby_connections(17721): onPayloadTransferUpdate
I/flutter (17721): Tt3c: -4780165049267211412:IMG_20210409_181224.jpg
I/flutter (17721): NoSuchMethodError: The method 'exists' was called on null.
I/flutter (17721): Receiver: null
I/flutter (17721): Tried calling: exists()
I/flutter (17721): InProgress
D/nearby_connections(17721): onPayloadTransferUpdate
I/flutter (17721): Success
W/NearbyConnections(17721): Failed to create Payload from ParcelablePayload: Java file not found at /storage/emulated/0/Download/.nearby/-4780165049267211412
W/NearbyConnections(17721): java.io.FileNotFoundException: open failed: EACCES (Permission denied)
W/NearbyConnections(17721): 	at android.os.ParcelFileDescriptor.openInternal(ParcelFileDescriptor.java:344)
W/NearbyConnections(17721): 	at android.os.ParcelFileDescriptor.open(ParcelFileDescriptor.java:231)
W/NearbyConnections(17721): 	at com.google.android.gms.nearby.connection.Payload$File.zza(Unknown Source:12)
W/NearbyConnections(17721): 	at com.google.android.gms.internal.nearby.zzfl.zza(Unknown Source:81)
W/NearbyConnections(17721): 	at com.google.android.gms.internal.nearby.zzav.zza(Unknown Source:5)
W/NearbyConnections(17721): 	at com.google.android.gms.internal.nearby.zzdx.dispatchTransaction(Unknown Source:28)
W/NearbyConnections(17721): 	at com.google.android.gms.internal.nearby.zzb.onTransact(Unknown Source:22)
W/NearbyConnections(17721): 	at android.os.Binder.execTransactInternal(Binder.java:1159)
W/NearbyConnections(17721): 	at android.os.Binder.execTransact(Binder.java:1123)
D/nearby_connections(17721): onPayloadReceived
D/AndroidRuntime(17721): Shutting down VM
E/AndroidRuntime(17721): FATAL EXCEPTION: main
</details>

**Comment by mannprerak2** [comment]
Did you request the Read files permission on runtime?

**Comment by virajmehta7** [comment]
Got same issue on Android 10

**Comment by mannprerak2** [comment]
I am able to replicate this on Android 11. 

<details>
<summary>Crash Logs</summary>

```
W/NearbyConnections(19143): Failed to create Payload from ParcelablePayload: Java file not found at /storage/emulated/0/Download/.nearby/-5722419836622414852
W/NearbyConnections(19143): java.io.FileNotFoundException: open failed: EACCES (Permission denied)
W/NearbyConnections(19143):     at android.os.ParcelFileDescriptor.openInternal(ParcelFileDescriptor.java:344)
W/NearbyConnections(19143):     at android.os.ParcelFileDescriptor.open(ParcelFileDescriptor.java:231)
W/NearbyConnections(19143):     at com.google.android.gms.nearby.connection.Payload$File.zza(Unknown Source:12)
W/NearbyConnections(19143):     at com.google.android.gms.internal.nearby.zzfl.zza(Unknown Source:81)
W/NearbyConnections(19143):     at com.google.android.gms.internal.nearby.zzav.zza(Unknown Source:5)
W/NearbyConnections(19143):     at com.google.android.gms.internal.nearby.zzdx.dispatchTransaction(Unknown Source:28)
W/NearbyConnections(19143):     at com.google.android.gms.internal.nearby.zzb.onTransact(Unknown Source:22)
W/NearbyConnections(19143):     at android.os.Binder.execTransactInternal(Binder.java:1170)
W/NearbyConnections(19143):     at android.os.Binder.execTransact(Binder.java:1134)
D/nearby_connections(19143): onPayloadReceived
D/AndroidRuntime(19143): Shutting down VM
E/AndroidRuntime(19143): FATAL EXCEPTION: main
E/AndroidRuntime(19143): Process: com.pkmnapps.nearby_connections_example, PID: 19143
E/AndroidRuntime(19143): java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.io.File.getAbsolutePath()' on a null object reference
E/AndroidRuntime(19143):        at com.pkmnapps.nearby_connections.NearbyConnectionsPlugin$13.onPayloadReceived(NearbyConnectionsPlugin.java:421)
E/AndroidRuntime(19143):        at com.google.android.gms.internal.nearby.zzaw.notifyListener(Unknown Source:10)
E/AndroidRuntime(19143):        at com.google.android.gms.common.api.internal.ListenerHolder.notifyListenerInternal(com.google.android.gms:play-services-base@@17.3.0:18)
E/AndroidRuntime(19143):        at com.google.android.gms.common.api.internal.ListenerHolder$zaa.handleMessage(com.google.android.gms:play-services-base@@17.3.0:6)
E/AndroidRuntime(19143):        at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime(19143):        at com.google.android.gms.internal.base.zap.dispatchMessage(com.google.android.gms:play-services-base@@17.3.0:8)
E/AndroidRuntime(19143):        at android.os.Looper.loop(Looper.java:245)
E/AndroidRuntime(19143):        at android.app.ActivityThread.main(ActivityThread.java:8004)
E/AndroidRuntime(19143):        at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(19143):        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:631)
E/AndroidRuntime(19143):        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:978)
I/Process (19143): Sending signal. PID: 19143 SIG: 9
```
</details>

**Comment by GauravGunjal43** [comment]
same issue, app crashing on receivers side, tried reversing the roles but still it crashes  on android 10 devices

**Comment by mannprerak2** [comment]
Glad to see there are so many users 😆 

Since Android 10, they have changed the storage access rules and we can't get the file path using my current implementation.

**For android 10**: you can add this attribute to your application element in your AndroidManifest.xml
```
android:requestLegacyExternalStorage="true" 
```

**For android 11 and beyond**: Seems like I'll have update the plugin code. I'll try to make a fix soon

**Comment by mannprerak2** [comment]
This is blocked by the underlying android nearby connections SDK. See this issue: https://github.com/android/connectivity-samples/issues/176

**Comment by GauravGunjal43** [comment]
> Glad to see there are so many users 😄
> 
> Since Android 10, they have changed the storage access rules and we can't get the file path using my current implementation.
> 
> **For android 10**: you can add this attribute to your application element in your AndroidManifest.xml (just a quick fix until I fix the plugin)
> 
> ```
> <application ...
> android:requestLegacyExternalStorage="true" 
> ...>
> ```
> 
> **For android 11 and beyond**: Seems like I'll have update the plugin code. I'll try to make a fix soon

this works but the file is received in a hidden folder named '.nearby' in the downloads directory. How do i change the directory of received files because it is not accessible.

**Comment by GauravGunjal43** [comment]
any updates on this?

**Comment by mannprerak2** [comment]
We can't do anything until they fix this in the underlying android nearby connections library.

**Comment by mannprerak2** [comment]
Hi everyone, I have released a new version with a fix for this - https://pub.dev/packages/nearby_connections/versions/3.1.0

- `Payload.uri` is now exposed and is usable on all android versions.
- The files are stored under `DOWNLOADS/.nearby/` which is a hidden folder and the file cannot be renamed in-place on Android 11 and beyond. So you must move and rename this file.
- A convenience method `copyFileAndDeleteOriginal` has been added to work with `uri` since they cannot be used directly.
- `Payload.filepath` has been deprecated and is only available on Android 10 and below.
